### ISSUE TITLE
Bug: Filters are executed even when controller does not exist with Auto Routing (Legacy)

### ISSUE THREAD

**Issue by kenjis** [codeigniter4/CodeIgniter4#7205]
### PHP Version

8.1

### CodeIgniter4 Version

4.3.1 and develop ea1b62872329076616287a20b879efb628c35d1a

### CodeIgniter4 Installation Method

Composer (using `codeigniter4/appstarter`)

### Which operating systems have you tested for this bug?

macOS

### Which server did you use?

cli-server (PHP built-in webserver)

### Database

n/a

### What happened?

When enabling Auto Routing (Legacy), if we navigate to the URI that does not exist,
global filters are executed.

### Steps to Reproduce

```diff
--- a/app/Config/Filters.php
+++ b/app/Config/Filters.php
@@ -31,7 +31,7 @@ class Filters extends BaseConfig
         'before' => [
             // 'honeypot',
             // 'csrf',
-            // 'invalidchars',
+            'invalidchars',
         ],
         'after' => [
             'toolbar',
```
```diff
--- a/app/Config/Routes.php
+++ b/app/Config/Routes.php
@@ -19,7 +19,7 @@ $routes->set404Override();
 // where controller filters or CSRF protection are bypassed.
 // If you don't want to define all routes, please use the Auto Routing (Improved).
 // Set `$autoRoutesImproved` to true in `app/Config/Feature.php` and set the following to true.
-// $routes->setAutoRoute(false);
+$routes->setAutoRoute(true);
 
 /*
  * --------------------------------------------------------------------
```
Navigate to `http://localhost:8080/xxx`.

The stack trace:
```
InvalidChars.php:52, CodeIgniter\Filters\InvalidChars->before()
Filters.php:173, CodeIgniter\Filters\Filters->run()
CodeIgniter.php:473, CodeIgniter\CodeIgniter->handleRequest()
CodeIgniter.php:368, CodeIgniter\CodeIgniter->run()
index.php:67, require_once()
rewrite.php:47, {main}()
```

### Expected Output

No filter is executed.

### Anything else?

But if we navigate to `http://localhost:8080/404`, no filter is executed.
Because `404` is invalid as a classname, and throw an exception here:
https://github.com/codeigniter4/CodeIgniter4/blob/ea1b62872329076616287a20b879efb628c35d1a/system/Router/AutoRouter.php#L98-L100

**Comment by ping-yee** [comment]
May I ask a question that how can I get the stack trace?

**Comment by kenjis** [comment]
I got the stack trace with PhpStrom step debugging by Xdebug.
https://www.jetbrains.com/help/phpstorm/stepping-through-the-program.html

**Comment by neznaika0** [comment]
Example of filter behavior for 404
https://forum.codeigniter.com/showthread.php?tid=86691
When a response is created via an exception, no filters are applied. It is ok. But it is not suitable for changing behavior for the API

**Comment by kenjis** [comment]
Closed by #7925
